stdlib/public/runtime/Casting.cpp

```diff
@@ -2755,6 +2755,50 @@ swift::swift_conformsToProtocol(const Metadata *type,
   goto recur;
 }
 
+/// Return the type metadata for a given type name.
+///
+/// This is a temporary implementation to support some Foundation use cases;
+/// it will only lookup a type that directly implements a protocol as only
+/// those types will have conformance table entries.
+static const Metadata *
+_metadataForName(const std::string &typeName) {
+  auto &C = Conformances.get();
+  const Metadata *foundMetadata = nullptr;
+
+  installCallbacksToInspectDylib();
+
+  pthread_mutex_lock(&C.SectionsToScanLock);
+
+  unsigned sectionIdx = 0;
+  unsigned endSectionIdx = C.SectionsToScan.size();
+
+  // TODO: implement typeName->metadata cache
+  // TODO: support looking up types for which there is no conformance entry
+  for (; sectionIdx < endSectionIdx; ++sectionIdx) {
+    auto &section = C.SectionsToScan[sectionIdx];
+    for (const auto &record : section) {
+      if (auto metadata = record.getCanonicalTypeMetadata()) {
+        std::string N = nameForMetadata(metadata, true);
+        if (N == typeName) {
+          foundMetadata = metadata;
+          break;
+        }
+      }
+    }
+  }
+
+  pthread_mutex_unlock(&C.SectionsToScanLock);
+
+  return foundMetadata;
+}
+
+extern "C"
+const Metadata *
+swift_getTypeByName(const char *typeName, size_t typeNameLength) {
+    std::string name(typeName, typeNameLength);
```

Comment:
You can pass an `llvm::StringRef` down instead of copying the string into a `std::string` here.